File: backend/app/schemas/modeling.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel
from app.models.modeling import ScenarioType, Granularity, LineItemType


# --- Scenario ---
class ScenarioCreate(BaseModel):
    name: str
    scenario_type: ScenarioType


class ScenarioOut(BaseModel):
    id: int
    model_id: int
    name: str
    scenario_type: ScenarioType
    created_at: datetime

    class Config:
        from_attributes = True


# --- TimePeriod ---
class TimePeriodCreate(BaseModel):
    label: str
    start_date: datetime
    end_date: datetime
    granularity: Granularity = Granularity.month
    sort_order: int


class TimePeriodOut(BaseModel):
    id: int
    model_id: int
    label: str
    start_date: datetime
    end_date: datetime
    granularity: Granularity
    sort_order: int

    class Config:
        from_attributes = True


# --- LineItem ---
class LineItemCreate(BaseModel):
    name: str
    section: Optional[str] = None
    item_type: LineItemType = LineItemType.input
    formula: Optional[str] = None
    account_id: Optional[int] = None
    sort_order: int = 0


class LineItemUpdate(BaseModel):
    name: Optional[str] = None
    section: Optional[str] = None
    item_type: Optional[LineItemType] = None
    formula: Optional[str] = None
    account_id: Optional[int] = None
    sort_order: Optional[int] = None


class LineItemOut(BaseModel):
    id: int
    model_id: int
    name: str
    section: Optional[str]
    item_type: LineItemType
    formula: Optional[str]
    account_id: Optional[int]
    sort_order: int

    class Config:
        from_attributes = True


# --- Model ---
class ModelCreate(BaseModel):
    name: str
    description: Optional[str] = None
    organisation_id: int


class ModelOut(BaseModel):
    id: int
    name: str
    description: Optional[str]
    organisation_id: int
    created_at: datetime
    scenarios: list[ScenarioOut] = []

    class Config:
        from_attributes = True


# --- CellValue ---
class CellValueUpsert(BaseModel):
    line_item_id: int
    scenario_id: int
    time_period_id: int
    value: float


class CellValueOverrideClear(BaseModel):
    line_item_id: int
    scenario_id: int
    time_period_id: int


# --- Grid ---
class GridCell(BaseModel):
    line_item_id: int
    time_period_id: int
    scenario_id: int
    value: Optional[float]
    is_formula: bool
    is_override: bool
    formula_text: Optional[str]
    error_message: Optional[str]


class GridResponse(BaseModel):
    model_id: int
    scenario_id: int
    time_periods: list[TimePeriodOut]
    line_items: list[LineItemOut]
    cells: list[GridCell]

File: frontend/src/services/modeling.ts
import axios from 'axios'
import type {
  Model, Scenario, TimePeriod, LineItem, GridResponse,
} from '@/types/modeling'

const api = axios.create({ baseURL: '/api' })

export const models = {
  list: (orgId: number) =>
    api.get<Model[]>('/models/', { params: { organisation_id: orgId } }).then(r => r.data),
  create: (payload: { name: string; description?: string; organisation_id: number }) =>
    api.post<Model>('/models/', payload).then(r => r.data),
  get: (id: number) =>
    api.get<Model>(`/models/${id}`).then(r => r.data),
  delete: (id: number) =>
    api.delete(`/models/${id}`),
}

export const scenarios = {
  list: (modelId: number) =>
    api.get<Scenario[]>(`/models/${modelId}/scenarios/`).then(r => r.data),
  create: (modelId: number, payload: { name: string; scenario_type: string }) =>
    api.post<Scenario>(`/models/${modelId}/scenarios/`, payload).then(r => r.data),
  delete: (modelId: number, scenarioId: number) =>
    api.delete(`/models/${modelId}/scenarios/${scenarioId}`),
}

export const timePeriods = {
  list: (modelId: number) =>
    api.get<TimePeriod[]>(`/models/${modelId}/time_periods/`).then(r => r.data),
  create: (modelId: number, payload: Omit<TimePeriod, 'id' | 'model_id'>) =>
    api.post<TimePeriod>(`/models/${modelId}/time_periods/`, payload).then(r => r.data),
  delete: (modelId: number, periodId: number) =>
    api.delete(`/models/${modelId}/time_periods/${periodId}`),
}

export const lineItems = {
  list: (modelId: number) =>
    api.get<LineItem[]>(`/models/${modelId}/line_items/`).then(r => r.data),
  create: (modelId: number, payload: Partial<LineItem>) =>
    api.post<LineItem>(`/models/${modelId}/line_items/`, payload).then(r => r.data),
  update: (modelId: number, itemId: number, payload: Partial<LineItem>) =>
    api.patch<LineItem>(`/models/${modelId}/line_items/${itemId}`, payload).then(r => r.data),
  delete: (modelId: number, itemId: number) =>
    api.delete(`/models/${modelId}/line_items/${itemId}`),
}

export const grid = {
  calculate: (modelId: number, scenarioId: number) =>
    api.post<GridResponse>(`/models/${modelId}/calculate`, null, {
      params: { scenario_id: scenarioId },
    }).then(r => r.data),
  get: (modelId: number, scenarioId: number) =>
    api.get<GridResponse>(`/models/${modelId}/grid`, {
      params: { scenario_id: scenarioId },
    }).then(r => r.data),
  updateCell: (modelId: number, payload: {
    line_item_id: number
    scenario_id: number
    time_period_id: number
    value: number
  }) => api.patch(`/models/${modelId}/cells/`, payload).then(r => r.data),
  clearOverride: (modelId: number, payload: {
    line_item_id: number
    scenario_id: number
    time_period_id: number
  }) => api.delete(`/models/${modelId}/cells/override`, { data: payload }).then(r => r.data),
}
